Expanded the description how to size the memory of the JVM (#15210)
elastic/logstash: docs/static/config-details.asciidoc

* all other lines are rejected

[[memory-size]]
==== Setting the memory size

The memory of the JVM executing {ls} can be divided in two zones: heap and off-heap memory.

> In the heap refers to Java heap, which contains all the Java objects created by {ls} during it's operation, see <<heap-size>> for

Reviewer: ```suggestion
In the heap refers to Java heap, which contains all the Java objects created by {ls} during its operation, see <<heap-size>> for
```

description on how to size it.

> What's not part of the heap is named off-heap and consist in memory that can be used and controlled by {ls}, generally

Reviewer: ```suggestion
What's not part of the heap is named off-heap and consists of memory that can be used and controlled by {ls}, generally
```

value to prevent the heap from resizing at runtime, which is a very costly
process.

* If you plan to use some plugins that leverages direct memory, like Beats, TCP or HTTP input,
 set the direct memory size (-XX:MaxDirectMemorySize) accordingly to the scale of load

>  they need to sustain.

Author: > Do we expect users to know which plugins use direct memory?

No we don't expect. For the user it's difficult to understand which plugin uses direct memory because it's something like an internal implementation "detail" however this impacts directly the user itself. I think that the best we can provide is a list of the plugins that uses direct memory. The user, when doing memory sizing has to be aware of this, because could impact him. For example by default users tend to think that all the memory used by Logstash is the JavaHeap, but that's not true, the JVM could consume other percentage of memory that's not immediately evident.
Just as an example, by default JVM size the direct memory space to the same size of Java heap size (Xmx, the maximum heap) this means that if a user has an host of 8GB, he could think to be safe (avoid the kernel memory killer or avoid disk swapping) setting Java heap to 6Gb (he leave 2Gb for the OS), but that's not true. Because the JVM can potentially consume 6 (heap) + 6 (direct memory) + 30% of heap for its internal structures, so 14GB.

> Also, will a typical user know how much load a plugin will be asked to handle when they are sizing their JVM? Or is something they'll likely be tweaking after OOM problems.

Generally yes, when hits OOM he fix the size, but knowing upfront the load he can size correctly the host and the JVM to sustain that load without risking OOM or hit some performance issues due to swapping.

value to prevent the heap from resizing at runtime, which is a very costly
process.

* Some input plugins, such as Elastic Agent, Beats, TCP, and HTTP inputs, use direct memory, allocated outside the JVM heap.

> Set the direct memory size (-XX:MaxDirectMemorySize) to accommodate the load you expect these plugins to handle.

Reviewer: We can move this into the off-heap memory section

`jmap` command line utility distributed with Java or by using VisualVM. For more
info, see <<profiling-the-heap>>.
// end::heap-size-tips[]

[[off-heap-size]]
===== Setting the off-heap size

The operating system, persistent queue mmap pages, direct memory, and other processes require memory in addition to memory allocated to heap size.

Internal JVM data structures, thread stacks, memory mapped files and direct memory for input/output (IO) operations are all parts of the off-heap JVM memory.
Thread stacks zone contains the list of stack frames for each Java thread spinned by the JVM; each frame keeps the local arguments passed during method calls.
Read on <<stacks-size>> if the size needs to be adapted to the processing needs.
Memory mapped files are not part of the Logstash's process off-heap memory, but consume RAM when paging files from disk.
These mapped files speed up the access to Persistent Queues pages, a performance improvement - or trade off - to reduce expensive disk operations such as read, write, and seek.
Some network I/O operations also resort to in-process direct memory usage to avoid, for example, copying of buffers between network sockets.

> Keep these memory requirements in mind as you calculate your ideal memory allocation.

Reviewer: ```suggestion
To summarize, we have 3 categories of memory usage, where 2 can be limited by the JVM and the other relies on available, free memory:

| Memory Type | Configured using | Used by |
| :---------------- | :------: | ----: |
| JVM Heap  |   -Xmx   | any normal object allocation |
| JVM direct memory |   -XX:MaxDirectMemorySize   | beats, tcp and http inputs |
| Native memory  |  N/A   | Persistent Queue Pages, Thread Stacks |

Keep these memory requirements in mind as you calculate your ideal memory allocation.
```

as the Operating System and other processes will require memory too.

For a {ls} instance with persistent queue (PQ) enabled on multiple pipelines, we could
estimate memory consumption using:

[source,text]
-----
pipelines number * (pipeline threads * stack size + 2 * PQ page size) + direct memory + Java heap
-----

NOTE: Each Persistent Queue requires that at least head and tail pages are present and accessible in memory.
The default page size is 64 MB so each PQ requires at least 128 MB of heap memory, which can be a significant source
of memory consumption per pipeline.

Another component of memory consumption is the thread stack size, allocated per thread.
Plugins, depending on their type (inputs, filters, and outputs), have different thread models.
Every input plugin runs in its own thread and can potentially spawn others. For example, each JDBC input
plugin launches a scheduler thread. Netty based plugins like TCP, Beats or HTTP input spawn a thread pool with 2 * number_of_cores threads.
Output plugins may also start helper threads, such as a connection management thread for each
{es} output instance.

> Every pipeline, then, has its own thread responsible to manage the pipeline lifecycle.

Reviewer: ```suggestion
Every pipeline, also, has its own thread responsible to manage the pipeline lifecycle.
```

> Every pipeline, then, has its own thread responsible to manage the pipeline lifecycle.

Reviewer: I'd suggest moving this explanation to the off heap section, then the sizing section can focus just on which variables to use and how to mix them together

Author: Fixed by 28e4c19b36250618f1ba0e32091d0c9a33fe3c20

==== Setting the off-heap size

The operating system, persistent queue mmap pages, direct memory, and other processes require memory in addition to memory allocated to heap size.

> Internal JVM data structures, thread stacks, memory mapped files, direct memory for input/output (IO) operations, and the Java heap used to execute Logstash all use JVM memory.

Reviewer 2: After tech review is complete, we can work together to refine final wording. You've got some good info here, and it deserves all the clarity we can give it. :-)

> Internal JVM data structures, thread stacks, memory mapped files, direct memory for input/output (IO) operations, and the Java heap used to execute Logstash all use JVM memory.

Reviewer: "all use JVM memory" maybe "use off-heap memory" instead? Many people will equate jvm memory to on-heap memory.

Author: Fixed by e497672c84733a3a52453307afece75e20e1cb4f

Thread stacks zone contains the list of stack frames for each Java thread spinned by the JVM; each frame keeps the local arguments passed during method calls.

> Memory mapped files is not part of direct IO memory, but contains part of RAM that's mapped for direct access to files, avoiding the common filesystem operations (read, write, seek and  append).

Reviewer: ```suggestion
Memory mapped files are not part of the Logstash's process off-heap memory, but consume RAM when paging files from disk.
These mapped files speed up the access to Persistent Queues pages, a performance improvement - or trade off - to reduce expensive disk operations such as read, write, and seek.
```

> These mapped files are used to speed up the access to Persistent Queues, in particular each PQ file can be mapped and unmapped from memory.

Reviewer: ```suggestion
```

Direct memory is instead that part of native memory which contains byte buffers to be exchanged with the networking part, and are used to avoid copying of buffers between network sockets.
All these different kind of memories fall under the term "native memory" or JVM's off-heap memory.

Keep the memory requirements for these processes in mind as you calculate your ideal memory allocation.

> Java heap can be controlled by Xmx and Xms settings.

Reviewer: This is under the `==== Setting the off-heap size` topic, so we should move it else where or remove it.

Author: Fixed by e497672c84733a3a52453307afece75e20e1cb4f

These mapped files speed up the access to Persistent Queues pages, a performance improvement - or trade off - to reduce expensive disk operations such as read, write, and seek.
Some network I/O operations also resort to in-process direct memory usage to avoid, for example, copying of buffers between network sockets,
in particular, some input plugins such as Elastic Agent, Beats, TCP, and HTTP inputs, use direct memory.

Another component of memory consumption is the thread stack size, allocated per thread.
Plugins, depending on their type (inputs, filters, and outputs), have different thread models.
Every input plugin runs in its own thread and can potentially spawn others. For example, each JDBC input
plugin launches a scheduler thread. Netty based plugins like TCP, Beats or HTTP input spawn a thread pool with 2 * number_of_cores threads.
Output plugins may also start helper threads, such as a connection management thread for each
{es} output instance.
Every pipeline, also, has its own thread responsible to manage the pipeline lifecycle.

To summarize, we have 3 categories of memory usage, where 2 can be limited by the JVM and the other relies on available, free memory:

> | Memory Type | Configured using | Used by |

Reviewer 2: ```suggestion
[cols="<,<,<",options="header",]
|=====
| Memory Type | Configured using | Used by
| JVM Heap  |   -Xmx   | any normal object allocation
| JVM direct memory |   -XX:MaxDirectMemorySize   | beats, tcp and http inputs
| Native memory  |  N/A   | Persistent Queue Pages, Thread Stacks
|=====
```

Reviewer 2: @andsel, this formatting should fix your table problem.  There are some other formatting and content issues that need to be fixed as well, but I know that you were particularly concerned about this table.

> | Memory Type | Configured using | Used by |
> | :---------------- | :------: | ----: |
> | JVM Heap  |   -Xmx   | any normal object allocation |
> | JVM direct memory |   -XX:MaxDirectMemorySize   | beats, tcp and http inputs |
> | Native memory  |  N/A   | Persistent Queue Pages, Thread Stacks |

Reviewer 2: ```suggestion
[cols="<,<,<",options="header",]
|=====
| Memory Type | Configured using | Used by
| JVM Heap  |   -Xmx   | any normal object allocation
| JVM direct memory |   -XX:MaxDirectMemorySize   | beats, tcp and http inputs
| Native memory  |  N/A   | Persistent Queue Pages, Thread Stacks
|=====
```

Reviewer 2: I tested locally, and here's how it looks with these changes.  Is this what you wanted to see? 

<img width="739" alt="Screen Shot 2023-09-22 at 10 58 34 AM" src="https://github.com/elastic/logstash/assets/35154725/771057d7-c4ee-49dc-8b3f-e5ffbf1539a9">

info, see <<profiling-the-heap>>.
// end::heap-size-tips[]

[[off-heap-size]]
===== Setting the off-heap size

The operating system, persistent queue mmap pages, direct memory, and other processes require memory in addition to memory allocated to heap size.

Internal JVM data structures, thread stacks, memory mapped files and direct memory for input/output (IO) operations are all parts of the off-heap JVM memory.

> Thread stacks zone contains the list of stack frames for each Java thread spinned by the JVM; each frame keeps the local arguments passed during method calls.

Reviewer: ```suggestion
The zone for Thread stacks contains the list of stack frames for each Java thread created by the JVM; each frame keeps the local arguments passed during method calls.
```

Read on <<stacks-size>> if the size needs to be adapted to the processing needs.
Memory mapped files are not part of the Logstash's process off-heap memory, but consume RAM when paging files from disk.
These mapped files speed up the access to Persistent Queues pages, a performance improvement - or trade off - to reduce expensive disk operations such as read, write, and seek.

> Some network I/O operations also resort to in-process direct memory usage to avoid, for example, copying of buffers between network sockets,
> in particular, some input plugins such as Elastic Agent, Beats, TCP, and HTTP inputs, use direct memory.

Reviewer: ```suggestion
Some network I/O operations also resort to in-process direct memory usage to avoid, for example, copying of buffers between network sockets. Input plugins such as Elastic Agent, Beats, TCP, and HTTP inputs, use direct memory.
```

> Another component of memory consumption is the thread stack size, allocated per thread.

Reviewer: We're back to talking about thread stacks, maybe we can move lines 96,97 here.

Author: Done with 1a990421ca669b33b6b533d99613bce1436694a6

By default, a JVM's off-heap direct memory limit is the same as the heap size. Check out <<plugins-inputs-beats-memory,beats input memory usage>>.

> Consider setting `-XX:MaxDirectMemorySize` to half of the heap size.

Reviewer: This entry should be moved down, and enhanced with some reasoning, since the new content introduces how to set it, so guidance on the values should come afterwards.

Author: Updated by commit d692dc1466c94a2d449108b50dc9e67d1b3483a4

Internal JVM data structures, thread stacks, memory mapped files, direct memory for input/output (IO) operations, and the Java heap used to execute Logstash all use JVM memory. 
Keep the memory requirements for these processes in mind as you calculate your ideal memory allocation.

> In addition to Java heap and direct memory, you must account for native memory as part of the overall memory consumption for {ls} processes.

Reviewer: this is the first place we mention these "three kinds of memories" but we don't explain, for example, direct vs native memory. maybe the line above can be slightly altered to provide this distinction (or just make the on-heap vs off heap distinction).

Author: Updated by commit d692dc1466c94a2d449108b50dc9e67d1b3483a4

Java heap can be controlled by Xmx and Xms settings. 
Direct memory space is controlled by the `-XX:MaxDirectMemorySize` setting. 
Total JVM memory allocation must be estimated and is controlled indirectly using Java heap and direct memory settings.

As you make your capacity calculations, keep in mind that the JVM can't consume more than 70% of the host memory available.

For a {ls} instance with persistent queue (PQ) enabled on multiple pipelines, we could
estimate memory consumption using:

[source,text]
-----
pipelines number * (workers * stack size + 2 * PQ page size) + direct memory + Java heap

> -----

Reviewer: maybe add a note here explaining one of the biggest memory hogs, "2* PQ page size", with something like "a PQ may page into memory up to two queue pages".

Author: Fixed by b352be45ace4811aa55057187abe3a572f933e4f

By default, a JVM's off-heap direct memory limit is the same as the heap size. Check out <<plugins-inputs-beats-memory,beats input memory usage>>.
Consider setting `-XX:MaxDirectMemorySize` to half of the heap size.

The total amount of memory used by the JVM comprises native memory allocated for internal JVM data structures, thread stacks,
memory mapped files, the direct memory used for IO operations and the Java heap used to execute Logstash.

> For a Logstash instance running a certain number of pipelines, all with PQ enabled we could
> estimate a memory consumption as follow:
> +
> [source,text]
> -------------------------------------
> pipelines number * (workers * stack size + 2 * PQ page size) + direct memory + Java heap
> -------------------------------------
> 
> As an example, considering a Logstash instance running 10 pipelines and that JVM by default allocates as much direct memory as Java heap,
> we could end with:
> * native memory: 10 * (12 * 1Mb + 128Mb), 1.4Gb
> * direct memory: 4Gb
> * Java heap: 4Gb
> There is an amount of native memory, other than Java heap and direct memory, that has to be accounted to size
> the overall memory consumption of Logstash process and that could only be estimated and doesn't have a direct control.
> In general the total memory consumed by a JVM running Logstash can only be estimated. Java heap can be controlled
> by Xmx and Xms settings, direct memory space is controlled by the `-XX:MaxDirectMemorySize` setting; but a general rule
> the JVM can't consume more than 70% of the host available memory, so some capacity calculations has to be done to avoid
> memory pages get swapped on disk.

Reviewer 2: ```suggestion
In addition to Java heap and direct memory, you must account for native memory as part of the overall memory consumption for {ls} processes.

Java heap can be controlled by Xmx and Xms settings. 
Direct memory space is controlled by the `-XX:MaxDirectMemorySize` setting. 
Total JVM memory allocation must be estimated and is controlled indirectly using Java heap and direct memory settings.

As you make your capacity calculations, keep in mind that the JVM can't consume more than 70% of the host memory available.

For a {ls} instance with persistent queue (PQ) enabled on multiple pipelines, we could
estimate memory consumption using:

[source,text]
-----
pipelines number * (workers * stack size + 2 * PQ page size) + direct memory + Java heap
-----

**Example**

Consider a {ls} instance running 10 pipelines.
Keep in mind that, by default, JVM allocates direct memory equal to memory allocated for Java heap.

The calculation results in:

* native memory: 1.4Gb  [derived from 10 * (12 * 1Mb + 128Mb)]
* direct memory: 4Gb
* Java heap: 4Gb
```

pipelines number * (workers * stack size + 2 * PQ page size) + direct memory + Java heap
-------------------------------------

As an example, considering a Logstash instance running 10 pipelines and that JVM by default allocates as much direct memory as Java heap,
we could end with:

> * native memory: 10 * (12 * 1Mb + 128Mb), 1.4Gb

Author: Should we exaplain how to count the threads in various context, or for the example we should be more specific, saying that "considering a single pipeline with such input and these other output we could count the threads up to...)". How can we explain how to count the threads? OR should we drop altogether  the section?

> * native memory: 10 * (12 * 1Mb + 128Mb), 1.4Gb

Author: I tried to better extend the description with commit a62b68a89f6690b15c190e8107655ed382b99113
